### ISSUE TITLE
`restat` is broken when rebuilding build.ninja in 1.6.0

### ISSUE THREAD

**Issue by jimporter** [ninja-build/ninja#987]
I updated to ninja 1.6.0 and found that `restat` commands produce an infinite loop. I didn't see anything in the release notes to suggest that `restat` had changed, but if I'm doing something wrong, let me know.

Below is a sample build.ninja that reproduces the issue. I don't think the `PHONY` part is actually necessary, but I included it just in case, since that's necessary for the real-world scenario this is used in.

PS: Sorry for missing the call for testing before 1.6.0 was released!

```
rule rescan
  command = echo rescanning...
  generator = 1
  restat = 1

rule regenerate
  command = echo regenerating...
  generator = 1

build PHONY: phony
build .bfg_watch: rescan | PHONY
build build.ninja: regenerate | .bfg_watch
```

**Comment by nico** [comment]
I'd guess this is "generators can now run more often than twice if needed (#908)" from the release notes. Your generator probably doesn't write everything it promises to write, and ninja now runs generator rules until they're clean (see that issue and the mail thread linked from there) this bug is exposed.

I'm guessing you can fix things by tweaking your generator.

**Comment by martine** [comment]
Perhaps we should implicitly restat the generator outputs and die if the command didn't regenerate them or something?

**Comment by jimporter** [comment]
I tried removing the `generator` attribute from the rescan rule, but it didn't change anything. In 1.5.3, what would happen is it would run the rescan rule once, see that nothing had changed, and then continue on with the build. Specifically, this means that I was never (to my knowledge) making the generators run twice.

The generator I'm using for this is my own project, [bfg9000](https://github.com/jimporter/bfg9000). The general idea is that the rescan rule checks to see if any glob patterns in the source build file have changed, and if so, to force the regenerate rule to run and rebuild the build.ninja file. Normally, none of the globs have changed, so it should run the rescan and then recognize that it doesn't need to do anything else.

**Comment by jimporter** [comment]
Oh, and the main reason I was using `generator = 1` for the rescan rule (which doesn't actually generate a build file itself) was because I wanted to ensure that it didn't get removed when running `ninja -t clean`. It's possible I don't need that at all, but perhaps it would be useful to have a way to say "don't clean things made by this rule".

**Comment by jimporter** [comment]
> Specifically, this means that I was never (to my knowledge) making the generators run twice.

Looking at the patch from #908 more, this may have been what's happening. All I really need is a way to say "automatically regenerate the build file if this arbitrary command says that something is out-of-date". It'd be easiest if I could do that by having said command touch a file, since that way works with GNU Make as well. If it helps, here's the issue where I planned this out for my build system: https://github.com/jimporter/bfg9000/issues/1

**Comment by nicolasdespres** [comment]
On Thu, Jul 9, 2015 at 10:10 PM, Jim Porter <notifications@github.com>
wrote:

> Oh, and the main reason I was using generator = 1 for the rescan rule
> (which doesn't actually generate a build file itself) was because I wanted
> to ensure that it didn't get removed when running ninja -t clean. It's
> possible I don't need that at all, but perhaps it would be useful to have a
> way to say "don't clean things made by this rule".
>
Your rescan command could create a hardlink from .bfg_watch to
.bfg_watch_not_cleanedup
when it has finished its job. At start up if .bfg_watch does not exist it
could create a hardlink from .bfg_watch_not_cleanedup to .bfg_watch if
.bfg_watch_not_cleanedup exists, then proceed normally. Obviously,
.bfg_watch_not_cleanedup must not appear in the output list of the rescan
build edge. I know it is just a workaround, but right now I can not figure
out an easy way to extend the clean tool to help you in your task.

Cheers,

-- 
Nicolas Desprès

**Comment by nico** [comment]
Would you mind if I added bfg9000 to https://github.com/martine/ninja/wiki/List-of-generators-producing-ninja-build-files ?

For your problem: @nicolasdespres's suggestion is the best I can think of too. If you don't want to rely on hard links, your rescan could write the list of globs and matching files to a different file not mentioned in the ninja build manifest, and make .bfg_watch an empty file that's only touched if the secret file on the side changes.

Out of interest, why do you want .bfg_watch not to be deleted by -t clean?

**Comment by jimporter** [comment]
> Would you mind if I added bfg9000 to https://github.com/martine/ninja/wiki/List-of-generators-producing-ninja-build-files ?

Go for it! It's not quite in a widely-usable state (Windows support is spotty), but I'm getting pretty close to a release.

> For your problem: @nicolasdespres's suggestion is the best I can think of too.

It sounds like that only applies to the issue of preserving `.bfg_watch` from `ninja -t clean`. I don't quite see how it would fix the (almost-) infinite regeneration loop I get stuck in. Am I just missing something?

> Out of interest, why do you want .bfg_watch not to be deleted by -t clean?

To be honest, I didn't think about it very hard in the first place. It might be ok if it gets deleted, but I'd have to figure out what actually happens then.

**Comment by nico** [comment]
>  I don't quite see how it would fix the (almost-) infinite regeneration loop I get stuck in.

I guess the problem is that if you have a rule that builds `build.ninja`, then this is executed as long as it's dirty (i.e. it's not the `generator` attribute as I had claimed above). And in your example, build.ninja depends on .bfg_watch, which depends on nothing, which means it's always dirty.

You want something that's always dirty, but only for one iteration of the generator. Hmmm…

(One thing that should work and that's likely faster than what you have now is to make the step that writes .bfg_watch have a depfile that lists all directories containing files matching glob patterns. Then ninja will use these as inputs and stat all of them at startup, and only rerun your rule if the glob patterns have changed.)

**Comment by jimporter** [comment]
> One thing that should work and that's likely faster than what you have now is to make the step that writes .bfg_watch have a depfile that lists all directories containing files matching glob patterns.

Unfortunately, the globs search recursively into subdirs, so the mtime of the root search directory isn't sufficient.

In any case, I figured it out. It's a combination of a bug in my `build.ninja` and a patch that got backed out of ninja. See https://github.com/martine/ninja/issues/874. If I reapply https://github.com/martine/ninja/pull/223 and add `restat = 1` to my `regenerate` rule, everything works as it should.

**Comment by nico** [comment]
> Unfortunately, the globs search recursively into subdirs, so the mtime of the root search directory isn't sufficient.

I mean all subdirs containing files matching globs (or just all subdirs if you have `**` style globs).

**Comment by jimporter** [comment]
Hmm, that could work. It could cause some unnecessary reconfigures though, e.g. if you add a .hpp file to a directory where you globbed for .cpp files. I guess it depends on whether more false positives would outstrip the performance benefits of not actually checking all the globs precisely...

**Comment by jimporter** [comment]
@nico: Ok, I did as you suggested and created a depfile when building the build.ninja files (ditto for Makefiles) and it seems to work, although I need to write some tests to be 100% sure. In the short term, that means I probably don't need this bug to be fixed, but I could see it being useful in the future as an optimization for bfg9000. With #223 re-landed, I could exit from the regeneration script early if it turns out that the globs *haven't* actually changed (e.g. if files were added to a globbed directory, but they didn't match the glob).

It's up to you whether you think this should remain open or not, but I'd lean (not strongly) towards leaving it open, since the behavior is slightly inconsistent and only exists to work around a crash, as far as I can tell from #874.

**Comment by nico** [comment]
We can't reland #223 as-is as it causes crashes.

What we could do is stop the generator loop if the first generator run doesn't touch build.ninja and the rule is marked restat. I think this would fix this bug without reintroducing the crashes. Let's leave this open.

**Comment by jimporter** [comment]
> We can't reland #223 as-is as it causes crashes.

Right, but I think the general idea in that patch is good. I'd guess it just reveals bustage somewhere else...

In any case, I've written sufficiently-thorough tests that I'm convinced that creating a depfile with the searched directories actually does what I need (although it does cause more regens than absolutely necessary), so I can use ninja 1.6.0 now. :)

**Comment by jimporter** [comment]
I realized one issue with using a depfile on the "regenerate build.ninja" rule: if you do something that creates a temp file (e.g. edit a file with emacs in its default configuration), that will mark build.ninja as dirty, even if the temp file is immediately removed. So, it'd be nice if this issue got fixed somehow. :)

(I might try to look at this when I've finished resolving the other major issues in my build system, but no promises.)
